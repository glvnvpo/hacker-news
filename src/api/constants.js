// @flow

type ID = number | string;

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = () => `${BASE}/newstories.json`;
export const ITEM = (id: ID)=> `${BASE}/item/${id}.json`;