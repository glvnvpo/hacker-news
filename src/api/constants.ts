// @flow

import {ID, URL} from '../types';

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = (): URL => `${BASE}/newstories.json`;
export const ITEM = (id: ID): URL => `${BASE}/item/${id}.json`;