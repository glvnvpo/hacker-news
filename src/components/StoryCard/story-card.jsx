// @flow

import React from "react";
import {Link} from "react-router-dom";
import {Card} from "react-bootstrap";
import {isEmpty} from "lodash";
import parse from "html-react-parser";
import './styles.scss';
import {Spinner} from "../Spinner";
import type {Story} from '../../types';
import {getDateFromTimestamp} from "../../helpers/get-date-from-timestamp";

type Props = {
    story: Story;
    isLoading?: boolean;
	asLink?: boolean;
	to?: string;
	fieldsToShow?: Array<string>;
}

export const StoryCard = ({isLoading=false, story, asLink=false, to, fieldsToShow=['all'], ...rest}: Props) => {

	let {title, score, by, time, url, text, descendants} = story;

	const shouldShowField = (key: string): boolean => {
		if (fieldsToShow.includes("all")) {
			return true;
		}

		return fieldsToShow.includes(key);
	};

	return (
		<Card
			className='story mt-10 mb-20'
			as={asLink ? Link : undefined}
			to={to}
			{...rest}
		>
			{
				isLoading ? <Spinner className="mt-20 mb-20" /> :
					!isEmpty(story) ?
						<Card.Body>
							<Card.Title className="color-orange">{title}</Card.Title>
							<Card.Subtitle className="mb-2 color-grey">{score} points | {by} | {getDateFromTimestamp(time)}</Card.Subtitle>
							{
								shouldShowField("url") &&
								<Card.Subtitle className="mb-2 color-grey">
									{url ? <a href={url} target="_blank" rel="noreferrer">Source</a> : 'No source'}
								</Card.Subtitle>
							}

							{
								shouldShowField("descendants") &&
								<Card.Subtitle className="mb-2 color-grey">Comment count: {descendants}</Card.Subtitle>
							}

							{
								shouldShowField("text") &&
								(text ?
									<Card.Text as='span' className="bold color-dark-grey">
										{parse(text)}
									</Card.Text>
									: <span className='bold color-dark-grey'>To read the text, visit the source</span>)
							}
						</Card.Body>
						: <h6 className="error mt-20 mb-20">Some troubles in loading story</h6>
			}
		</Card>
	);
};