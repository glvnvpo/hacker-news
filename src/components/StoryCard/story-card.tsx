// @flow

import React, {FC, HTMLAttributes} from 'react';
import {Link} from 'react-router-dom';
import {Card} from 'react-bootstrap';
import {isEmpty} from 'lodash';
import parse from 'html-react-parser';
import './styles.scss';
import {Spinner} from '../Spinner';
import {Story} from '../../types';
import {getDateFromTimestamp} from '../../helpers/get-date-from-timestamp';

export enum Fields {
	ALL= 'all',
	URL = 'url',
	TEXT = 'text'
}

type Props = {
    story: Story | undefined;
    isLoading?: boolean;
	asLink?: boolean;
	to?: string;
	extraFieldsToShow?: Array<Fields>;
}

export const StoryCard: FC<Props & HTMLAttributes<any>> = ({story, isLoading=false, asLink=false, to,
															   extraFieldsToShow=[Fields.ALL], ...rest}) => {

	let {title, score, by, time, url, text, descendants} = story || {};

	const shouldShowField = (key: Fields): boolean => {
		if (extraFieldsToShow.includes(Fields.ALL)) {
			return true;
		}

		return extraFieldsToShow.includes(key);
	};

	return (
		<Card
			className='story mt-10 mb-20'
			as={asLink ? Link : undefined}
			to={to || ''}
			{...rest}
		>
			{
				isLoading ? <Spinner className='mt-20 mb-20' /> :
					!isEmpty(story) ?
						<Card.Body>
							<Card.Title className='color-orange'>{title}</Card.Title>
							<Card.Subtitle className='mb-2 color-grey'>{score}&nbsp;points&nbsp;| {by}&nbsp;| {getDateFromTimestamp(time)}</Card.Subtitle>

							{
								shouldShowField(Fields.URL) &&
								<Card.Subtitle className='url mb-2 color-grey'>
									{url ? <a href={url} target='_blank' rel='noreferrer'>Source</a> : 'No source link available'}
								</Card.Subtitle>
							}

							<Card.Subtitle className='comments-count mb-2 color-grey'>Comments count: {descendants}</Card.Subtitle>

							{
								shouldShowField(Fields.TEXT) &&
								(text ?
									<Card.Text as='span' className='bold color-dark-grey'>
										{parse(text)}
									</Card.Text>
									: <span className='bold color-dark-grey'>To read the text, visit the source</span>)
							}
						</Card.Body>
						: <h6 className='error mt-20 mb-20'>Some troubles in loading story</h6>
			}
		</Card>
	);
};