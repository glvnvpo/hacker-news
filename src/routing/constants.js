export const ROOT_PATH                 = '/';
export const MAIN_PAGE_PATH            = '/hacker-news';
export const SINGLE_STORY_PAGE_PATH    = `${MAIN_PAGE_PATH}/:id`;