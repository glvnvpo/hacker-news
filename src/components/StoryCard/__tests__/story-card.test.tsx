import React from 'react';
import {mount} from 'enzyme';
import {BrowserRouter} from 'react-router-dom';
import {StoryCard} from '../index';
import {Fields} from '../story-card';
import {MAIN_PAGE_PATH} from '../../../routing/constants';

describe('StoryCard', () => {
	let wrapper;

	const story = {
		id: 123,
		title: 'someTitle',
		by: 'someUser',
		time: '1662661417',
		score: 2,
		descendants: 1,
		url: 'https://reactjs.org/',
		text: 'someText',
	};
    
	it('should render correct StoryCard', () => {
		wrapper = mount(<StoryCard story={story} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should render correct StoryCard as link', () => {
		const props = {
			story,
			asLink: true,
			to: `${MAIN_PAGE_PATH}/${story.id}`,
			extraFieldsToShow: []
		};
		wrapper = mount(
			<BrowserRouter>
				<StoryCard {...props} />
			</BrowserRouter>
		);
		expect(wrapper.find('Link').exists()).toBeTruthy();
	});

	it('should render without url', () => {
		const props = {
			story: {
				...story,
				url: undefined
			}
		};
		wrapper = mount(<StoryCard {...props} />);
		expect(wrapper.find('div.url').text()).toBe('No source link available');
	});

	it('should render without text', () => {
		const props = {
			story: {
				...story,
				text: undefined
			}
		};
		const expected = 'To read the text, visit the source';
		wrapper = mount(<StoryCard {...props} />);
		expect(wrapper.text()).toMatch(expected);
	});
	
	it('should render story with error', () => {
		wrapper = mount(<StoryCard story={undefined} />);
		const expected = 'Some troubles in loading story';
		expect(wrapper.text()).toBe(expected);
	});

	it('should render Spinner', () => {
		wrapper = mount(<StoryCard story={story} isLoading={true}  />);
		expect(wrapper.find('Spinner')).toHaveLength(1);
	});

	it('should render url and not text', () => {
		const props = {
			story,
			extraFieldsToShow: [Fields.URL]
		};
		wrapper = mount(<StoryCard {...props}  />);
		expect(wrapper.find('.url').exists()).toBeTruthy();
		expect(wrapper.find('.text').exists()).toBeFalsy();
	});
});