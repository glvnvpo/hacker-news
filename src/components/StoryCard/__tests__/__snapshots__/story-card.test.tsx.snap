// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard should render correct StoryCard 1`] = `
<StoryCard
  story={
    {
      "by": "someUser",
      "descendants": 1,
      "id": 123,
      "score": 2,
      "text": "someText",
      "time": "1662661417",
      "title": "someTitle",
      "url": "https://reactjs.org/",
    }
  }
>
  <Card
    body={false}
    className="story mt-10 mb-20"
    to=""
  >
    <div
      className="story mt-10 mb-20 card"
      to=""
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardTitle
            className="color-orange"
          >
            <ForwardRef
              className="color-orange card-title"
            >
              <div
                className="color-orange card-title h5"
              >
                someTitle
              </div>
            </ForwardRef>
          </CardTitle>
          <CardSubtitle
            className="mb-2 color-grey"
          >
            <ForwardRef
              className="mb-2 color-grey card-subtitle"
            >
              <div
                className="mb-2 color-grey card-subtitle h6"
              >
                2
                 points | 
                someUser
                 | 
                08.09.2022
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardSubtitle
            className="url mb-2 color-grey"
          >
            <ForwardRef
              className="url mb-2 color-grey card-subtitle"
            >
              <div
                className="url mb-2 color-grey card-subtitle h6"
              >
                <a
                  href="https://reactjs.org/"
                  rel="noreferrer"
                  target="_blank"
                >
                  Source
                </a>
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardSubtitle
            className="comments-count mb-2 color-grey"
          >
            <ForwardRef
              className="comments-count mb-2 color-grey card-subtitle"
            >
              <div
                className="comments-count mb-2 color-grey card-subtitle h6"
              >
                Comments count: 
                1
              </div>
            </ForwardRef>
          </CardSubtitle>
          <CardText
            as="span"
            className="bold color-dark-grey"
          >
            <span
              className="bold color-dark-grey card-text"
            >
              someText
            </span>
          </CardText>
        </div>
      </CardBody>
    </div>
  </Card>
</StoryCard>
`;
